site-src/geps/gep-1282.md

```diff
@@ -98,12 +98,302 @@ Of course, from the section above, you can see that there are many other feature
+
+In https://github.com/kubernetes-sigs/gateway-api/blob/main/apis/v1beta1/shared_types.go add a new type:
+```go
+// BackendCapabilities references a Kubernetes Service and defines any backend 
+// capabilities that Service requires.  
+//
+// +optional
+type BackendCapabilities struct {
+    // BackendName is the name of the backend service to which these backend 
+    // capabilities apply.
+    // +required
+    BackendName string `json:”backendName”`
+
+    // BackendType is the type of backend to which these backend 
+    // capabilities apply.
+    //
+    // +required
+    BackendType BackendTargetRef `json:”backendType”`
+
+	// Spec specifies the capabilities supported by this
+    // Service wrapper, if any.  The capabilities are predefined, and have two
+    // characteristics in common: they are tightly bound to the backend, 
+    // rather than being important at the route level, and the service owner
+    // (application developer in the Gateway API personas) should control 
+    // these settings, not the owner of the Gateway implementation (the 
+    // Cluster Operator in the Gateway API personas).
+    // 
+    // Support: Extended
+	//
+    // +required
+	Spec BackendCapabilitiesSpec `json:”spec”`
+
+	// Status represents the current state of the backend and its capabilities.
+    // +optional
+	Status *BackendCapabilitiesStatus `json:”status,omitempty”`
+}
+
+// BackendTargetReference identifies an API object to apply BackendCapabilities 
+// to. 
+// 
+type BackendTargetReference struct {
+	// Group is the group of the target resource.
+	Group Group `json:"group"`
+
+	// Kind is kind of the target resource.
+    // Valid values include:
+    // * "Service" (the default if no Kind is specified)
+    // * "ServiceImport"
+    //
+    // Invalid values include:
+    //
+    // * "invalid/kind" - "/" is an invalid character
+    // +required
+	Kind string `json:"kind"`
+
+	// Version is the version of the target resource.
+	Version ObjectName `json:"version"`
+
+	// Name is the name of the target resource.
+	Name ObjectName `json:"name"`
+}
+
+// BackendCapabilitiesSpec defines a group of specific properties held by its
+// affiliated BackendCapabilitiesService, and specified by the Service owner 
+// (typically the application developer).  Examples of properties that 
+// belong in the BackendCapabilitiesSpec include: TLS re-encryption, session
+// state, timeouts, retries, Websockets, HTTP/2, load balancer algorithm, 
+// healthcheck info, external traffic policy, connection limits, error page,
+// protocol switch, alerts, and others. 
+//
+// Support: Extended
+//
+type BackendCapabilitiesSpec struct {
+
+	// +optional
+	Reencrypt *TLSReencryptType `json:”reencrypt,omitempty”`
+
+	// +optional
+    Websocket *WebsocketType `json:”websocket,omitempty”`
+
+	// +optional
+    Protocol *ProtocolType `json:”protocol,omitempty”`
+}
+
+type ReencryptType struct {
+    Version *TLSVersion
```

Comment:
I think many backends support a range of TLS versions, it may be worth allowing room for that here. As a broader question, is this absolutely required? Could we live without it in a v1?

Reply:
I also think having only `TLSVersion` is bit ambiguous. Usually I see something like `MinTLSVersion`, `MaxTLSVersion` and also a list of `AllowedCipherSuites` (and so far, ciphers have no effect for TLSv1.3).  

Maybe for TLS clients, like we have here, these settings are bit "less visible" than for servers: servers immediately get caught when running TLS scanners towards them and users will request for configurability. In case of clients, the lack of configurability might go under the radar for a while. Regardless, having configurability at least for min/max would be very nice.

Reply:
Good points.  Will replace with min/max version. Optional, so if not specified then there is no check.  If anyone feels strongly that we don't need it, we can remove it.

Reply:
> (and so far, ciphers have no effect for TLSv1.3).

@tsaarni, what do you mean by this?